solr/core/src/java/org/apache/solr/search/MultiThreadedSearcher.java

```diff
@@ -0,0 +1,415 @@
+        throw (RuntimeException) ex.getCause().getCause();
+      } else {
+        throw ex;
+      }
+    }
+
+    ScoreMode scoreMode = SolrMultiCollectorManager.scoreMode(firstCollectors);
+
+    return new SearchResult(scoreMode, ret);
+  }
+
+  static boolean allowMT(DelegatingCollector postFilter, QueryCommand cmd, Query query) {
+    if (postFilter != null || cmd.getSegmentTerminateEarly() || cmd.getTimeAllowed() > 0) {
+      return false;
+    } else {
+      MTCollectorQueryCheck allowMT = new MTCollectorQueryCheck();
+      query.visit(allowMT);
+      return allowMT.allowed();
+    }
+  }
+
+  /**
+   * A {@link QueryVisitor} that recurses through the query tree, determining if all queries support
+   * multi-threaded collecting.
+   */
+  private static class MTCollectorQueryCheck extends QueryVisitor {
+
+    private QueryVisitor subVisitor = this;
+
+    private boolean allowMt(Query query) {
+      if (query instanceof RankQuery || query instanceof GraphQuery || query instanceof JoinQuery) {
+        return false;
+      }
+      return true;
+    }
+
+    @Override
+    public void consumeTerms(Query query, Term... terms) {
+      if (!allowMt(query)) {
+        subVisitor = EMPTY_VISITOR;
+      }
+    }
+
+    @Override
+    public void consumeTermsMatching(
+        Query query, String field, Supplier<ByteRunAutomaton> automaton) {
+      if (!allowMt(query)) {
+        subVisitor = EMPTY_VISITOR;
+      } else {
+        super.consumeTermsMatching(query, field, automaton);
+      }
+    }
+
+    @Override
+    public void visitLeaf(Query query) {
+      if (!allowMt(query)) {
+        subVisitor = EMPTY_VISITOR;
+      }
+    }
+
+    @Override
+    public QueryVisitor getSubVisitor(BooleanClause.Occur occur, Query parent) {
+      return subVisitor;
+    }
+
+    public boolean allowed() {
+      return subVisitor != EMPTY_VISITOR;
+    }
+  }
+
+  static class MaxScoreResult {
+    final float maxScore;
+
+    public MaxScoreResult(float maxScore) {
+      this.maxScore = maxScore;
+    }
+  }
+
+  static class FixedBitSetCollector extends SimpleCollector {
+    private final LinkedList<FixedBitSet> bitSets = new LinkedList<>();
+    private final LinkedList<Integer> skipWords = new LinkedList<>();
+    private final LinkedList<Integer> skipBits = new LinkedList<>();
+
+    FixedBitSetCollector() {}
+
+    @Override
+    protected void doSetNextReader(LeafReaderContext context) throws IOException {
+      this.bitSets.add(null); // lazy allocate when collecting document(s)
+      this.skipWords.add(context.docBase / 64);
+      this.skipBits.add(context.docBase % 64);
+    }
+
+    @Override
+    public void collect(int doc) throws IOException {
+      FixedBitSet bitSet = this.bitSets.getLast();
+      final int idx = this.skipBits.getLast() + doc;
+
+      final int numWords = FixedBitSet.bits2words(idx + 1); // +1 to ensure minimum 1 word
+
+      if (bitSet == null) {
```

Comment:
I'm thinking lazy allocation makes sense since for some segments maybe nothing will ever be collected ...

... however `+1` here for segment bitset sizing seems too basic ...

... didn't (yet) find a number-of-documents-in-a-segment value ... `maxDoc / numberOfLeaves` might a way to initially approximate but once that's exceeded the question of how to upsize arises again ...

... thoughts?